### PR TITLE
Fix `has_one` `enum` `where` queries

### LINKED ISSUE #25128: Can't use enum value in  `where` key with `joins` for has_one relation query.
### Steps to reproduce

I wrote following code, but it isn't work.

```ruby
# bundle exec rails -v
class User < ApplicationRecord
  has_one :address
end

class Address < ApplicationRecord
  belongs_to :user
  enum kind: %i(home work)
end

User.joins(:address).where(addresses: {kind: 'home'}).to_sql
  #=> "SELECT \"users\".* FROM \"users\" INNER JOIN \"addresses\" ON \"addresses\".\"user_id\" = \"users\".\"id\" WHERE \"addresses\".\"kind\" = 'home'"
```

But, followin code is work.

```ruby
class User < ApplicationRecord
  has_many :items
end

class Item < ApplicationRecord
  belongs_to :user
  enum state: %i(pendding published)
end

User.joins(:items).where(items: {state: 'pendding'}).to_sql
  #=> "SELECT \"users\".* FROM \"users\" INNER JOIN \"items\" ON \"items\".\"user_id\" = \"users\".\"id\" WHERE \"items\".\"state\" = 0"
```
### Expected behavior

```ruby
User.joins(:address).where(addresses: {kind: 'home'}).to_sql
  #=> "SELECT \"users\".* FROM \"users\" INNER JOIN \"addresses\" ON \"addresses\".\"user_id\" = \"users\".\"id\" WHERE \"addresses\".\"kind\" = 0"
```

### Actual behavior

```ruby
User.joins(:address).where(addresses: {kind: 'home'}).to_sql
  #=> "SELECT \"users\".* FROM \"users\" INNER JOIN \"addresses\" ON \"addresses\".\"user_id\" = \"users\".\"id\" WHERE \"addresses\".\"kind\" = 'home'"
```

### System configuration
**Rails version**:Rails 5.1.0.alpha
# 

**Ruby version**:ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-darwin15]

### PR DESCRIPTION
**Pull request by maclover7**
Fixes #25128

### CODE REVIEW

File: activerecord/lib/active_record/table_metadata.rb [hunk 1]
```diff
@@ -44,17 +44,21 @@ def associated_with?(association_name)
     def associated_table(table_name)
       return self if table_name == arel_table.name
 
-      association = klass._reflect_on_association(table_name)
-      if association && !association.polymorphic?
-        association_klass = association.klass
-        arel_table = association_klass.arel_table.alias(table_name)
+      # First check what is passed in, and see if an associationexists in
```
Comment (meinac) [inline activerecord/lib/active_record/table_metadata.rb]: need a space here: `associationexists`.

File: activerecord/lib/active_record/table_metadata.rb [hunk 2]
```diff
@@ -44,17 +44,21 @@ def associated_with?(association_name)
     def associated_table(table_name)
       return self if table_name == arel_table.name
 
-      association = klass._reflect_on_association(table_name)
```
Comment (sgrif) [inline activerecord/lib/active_record/table_metadata.rb]: Seems like this entire PR could be simplified to just changing this line to

```ruby
association = klass._reflect_on_association(table_name) || klass._reflect_on_association(table_name.singularize)
```

**Comment by maclover7** [comment]
👍 updated @sgrif

File: activerecord/lib/active_record/table_metadata.rb [hunk 3]
```diff
@@ -44,10 +44,14 @@ def associated_with?(association_name)
...
+      # Check and see if an association exists for the `table_name` passed in,
+      # and if not, pluralize the `table_name` and check again. This is useful
+      # in a situation where you are querying with `has_many` association.
+      association = klass._reflect_on_association(table_name) || klass._reflect_on_association(table_name.singularize)
+
       if association && !association.polymorphic?
         association_klass = association.klass
-        arel_table = association_klass.arel_table.alias(table_name)
+        arel_table = association_klass.arel_table.alias(table_name.to_s.pluralize)
```
Comment (sgrif) [inline activerecord/lib/active_record/table_metadata.rb]: The `pluralize` isn't needed

File: activerecord/lib/active_record/table_metadata.rb [hunk 4]
```diff
@@ -44,10 +44,14 @@ def associated_with?(association_name)
     def associated_table(table_name)
       return self if table_name == arel_table.name
 
-      association = klass._reflect_on_association(table_name)
+      # Check and see if an association exists for the `table_name` passed in,
+      # and if not, pluralize the `table_name` and check again. This is useful
+      # in a situation where you are querying with `has_many` association.
```
Comment (sgrif) [inline activerecord/lib/active_record/table_metadata.rb]: :scissors:

**Comment by maclover7** [comment]
👍 updated again @sgrif :)

**Comment by maclover7** [comment]
👍

**Pull request merged**